test/shed_functional/base/twilltestcase.py

```diff
@@ -261,12 +261,16 @@ def commit_and_push( self, repository, hgrepo, options, username, password ):
         commands.commit( ui.ui(), hgrepo, **options )
         try:
             commands.push( ui.ui(), hgrepo, dest=url )
-        except Abort as a:
-            message = a
-            if 'authorization failed' in message:
+        except Exception as e:
+            if str(e).find('Pushing to Tool Shed is disabled') != -1:
                 return False
             else:
-                raise
+                return self.__push_again( hgrepo, url )
+        return True
+
+    def __push_again( hgrepo, url ):
+        try:
+            commands.push( ui.ui(), hgrepo, dest=url )
         except Exception as e:
             if str(e).find('Pushing to Tool Shed is disabled') != -1:
                 return False
```

Comment:
Actually if you want to be consistent with `export_capsule()` method, that would be:
```
    try:
        return _commit_and_push(hgrepo, url)
    except Exception:
        return _commit_and_push(hgrepo, url)

def _commit_and_push(hgrepo, url):
    try:
        commands.push( ui.ui(), hgrepo, dest=url )
    except Exception as e:
        if str(e).find('Pushing to Tool Shed is disabled') != -1:
            return False
        else:
            raise
    return True
```
which is still not extendable to more than 2 retries, but at least avoids the code duplication.